File: central-assistant/identity.mjs
/**
 * Signed forwarded identity — §5.3 of CHATBOT-CENTRAL-SERVICE-PLAN.md, built day one.
 *
 * Token = b64url(JSON payload) + "." + b64url(HMAC-SHA256(payloadB64, key)).
 * Payload: { userId, userName, role, vesselId?, tenantDomain, tuid?, iat, exp }.
 * TTL is short (default 60 s) — the token is a per-request assertion, not a session.
 *
 * PRODUCTION MINT CONTRACT (Stage 4, module-side — recorded here so it cannot drift):
 * the mint endpoint lives in the MODULE backend and MUST source the role from the
 * module's REAL resolved identity — `req.rbac` (the forwarded SAILERP role) — NEVER
 * from the mock `req.user.role` fallback in middleware/auth.ts. A token minted from
 * the mock would launder its over-permission behind a valid signature. A mint request
 * whose rbac source is "none" must be refused, not defaulted.
 *
 * Stage 2 provides mint only as a local test utility (mint-token.mjs) plus this
 * library; the service itself only VERIFIES.
 */
import { createHmac, timingSafeEqual } from 'node:crypto';

/** Mint and verify run on DIFFERENT machines (module backend vs central service) —
 *  bounded clock-skew leeway is required, PROVEN on the pilot: the AI server runs
 *  62 s ahead of the build workstation, making 60 s tokens expired-on-arrival.
 *  Leeway bounds BOTH directions: expiry gets grace, and a token future-dated
 *  beyond the leeway is rejected (anti-pre-mint). */
const CLOCK_LEEWAY_SEC = parseInt(process.env.IDENTITY_CLOCK_LEEWAY_SEC || '90', 10);

const b64url = (buf) => Buffer.from(buf).toString('base64url');

export function signIdentity(identity, key, ttlSec = 60) {
  if (!key) throw new Error('signing key required');
  const now = Math.floor(Date.now() / 1000);
  const payload = { ...identity, iat: now, exp: now + ttlSec };
  const body = b64url(JSON.stringify(payload));
  const mac = createHmac('sha256', key).update(body).digest('base64url');
  return `${body}.${mac}`;
}

/** @returns {{ok:true, identity:object} | {ok:false, reason:'missing'|'malformed'|'bad-signature'|'expired'}} */
export function verifyIdentity(token, key) {
  if (!token || typeof token !== 'string') return { ok: false, reason: 'missing' };
  const dot = token.lastIndexOf('.');
  if (dot <= 0) return { ok: false, reason: 'malformed' };
  const body = token.slice(0, dot);
  const mac = token.slice(dot + 1);
  const expected = createHmac('sha256', key).update(body).digest();
  let given;
  try {
    given = Buffer.from(mac, 'base64url');
  } catch {
    return { ok: false, reason: 'malformed' };
  }
  if (given.length !== expected.length || !timingSafeEqual(given, expected)) {
    return { ok: false, reason: 'bad-signature' };
  }
  let identity;
  try {
    identity = JSON.parse(Buffer.from(body, 'base64url').toString('utf8'));
  } catch {
    return { ok: false, reason: 'malformed' };
  }
  const now = Math.floor(Date.now() / 1000);
  if (!identity.exp || identity.exp + CLOCK_LEEWAY_SEC < now) return { ok: false, reason: 'expired' };
  if (identity.iat && identity.iat - CLOCK_LEEWAY_SEC > now) return { ok: false, reason: 'future-dated' };
  if (!identity.userId || !identity.role) return { ok: false, reason: 'malformed' };
  return { ok: true, identity };
}

File: central-assistant-py/app/identity.py
"""
Signed forwarded identity — §5.3 of CHATBOT-CENTRAL-SERVICE-PLAN.md. Python port of
central-assistant/identity.mjs, BYTE-COMPATIBLE on the wire with the Node implementation
and with the module-side TS twin (server/modules/assistant-api/identityToken.ts):

  token   = b64url(JSON payload) + "." + b64url(HMAC-SHA256(payloadB64, key))
  payload = { userId, userName, role, vesselId?, tenantDomain, tuid?, iat, exp }

TTL is short (60 s) — a per-request assertion, not a session. Verification allows a
bounded clock-skew leeway BOTH ways (PROVEN 62 s skew between our own hosts): expiry
gets grace; a token future-dated beyond the leeway is rejected (anti-pre-mint).

The service only VERIFIES. Production minting is module-side and must source the role
from the module's real forwarded identity (req.rbac) — never a mock fallback.
"""
from __future__ import annotations

import base64
import hashlib
import hmac
import json
import time
from dataclasses import dataclass
from typing import Any


def _b64url_encode(raw: bytes) -> str:
    return base64.urlsafe_b64encode(raw).decode("ascii").rstrip("=")


def _b64url_decode(s: str) -> bytes:
    pad = "=" * (-len(s) % 4)
    return base64.urlsafe_b64decode(s + pad)


def sign_identity(identity: dict[str, Any], key: str, ttl_sec: int = 60) -> str:
    """Test/tooling mint — same bytes Node's signIdentity produces for the same input."""
    if not key:
        raise ValueError("signing key required")
    now = int(time.time())
    payload = {**identity, "iat": now, "exp": now + ttl_sec}
    # Node: JSON.stringify → no spaces, key order as inserted. Python json.dumps with
    # separators=(",", ":") and ensure_ascii=False matches for the plain-ASCII payloads used.
    body = _b64url_encode(json.dumps(payload, separators=(",", ":"), ensure_ascii=False).encode("utf-8"))
    mac = hmac.new(key.encode("utf-8"), body.encode("ascii"), hashlib.sha256).digest()
    return f"{body}.{_b64url_encode(mac)}"


@dataclass(frozen=True)
class Verified:
    ok: bool
    identity: dict[str, Any] | None = None
    reason: str | None = None  # missing | malformed | bad-signature | expired | future-dated


def verify_identity(token: str | None, key: str, clock_leeway_sec: int = 90) -> Verified:
    if not token or not isinstance(token, str):
        return Verified(False, reason="missing")
    dot = token.rfind(".")
    if dot <= 0:
        return Verified(False, reason="malformed")
    body, mac = token[:dot], token[dot + 1:]
    expected = hmac.new(key.encode("utf-8"), body.encode("ascii"), hashlib.sha256).digest()
    try:
        given = _b64url_decode(mac)
    except Exception:
        return Verified(False, reason="malformed")
    if len(given) != len(expected) or not hmac.compare_digest(given, expected):
        return Verified(False, reason="bad-signature")
    try:
        identity = json.loads(_b64url_decode(body).decode("utf-8"))
    except Exception:
        return Verified(False, reason="malformed")
    if not isinstance(identity, dict):
        return Verified(False, reason="malformed")
    now = int(time.time())
    exp = identity.get("exp")
    iat = identity.get("iat")
    if not isinstance(exp, (int, float)) or exp + clock_leeway_sec < now:
        return Verified(False, reason="expired")
    if isinstance(iat, (int, float)) and iat - clock_leeway_sec > now:
        return Verified(False, reason="future-dated")
    if not identity.get("userId") or not identity.get("role"):
        return Verified(False, reason="malformed")
    return Verified(True, identity=identity)
